plans/dht/test/find_providers.go

```diff
@@ -2,19 +2,100 @@ package test
 
 import (
 	"fmt"
+	"time"
 
+	"github.com/ipfs/go-cid"
+	ipfsUtil "github.com/ipfs/go-ipfs-util"
 	"github.com/ipfs/testground/sdk/runtime"
 )
 
 // FindProviders implements the Find Providers Test case
 func FindProviders(runenv *runtime.RunEnv) {
-	fmt.Printf("Not implemented yet")
+	// Test Parameters
+	var (
+		timeout     = time.Duration(runenv.IntParamD("timeout_secs", 60)) * time.Second
+		randomWalk  = runenv.BooleanParamD("random_walk", false)
+		bucketSize  = runenv.IntParamD("bucket_size", 20)
+		autoRefresh = runenv.BooleanParamD("auto_refresh", true)
+		// pProviding  = runenv.IntParamD("p_providing", 10)
+		// pResolving  = runenv.IntParamD("p_resolving", 10)
+		// pFailing    = runenv.IntParamD("p_failing", 10)
+	)
 
-	/// --- Warm Up
+	ctx, _, dht, _, err := SetUp(runenv, timeout, randomWalk, bucketSize, autoRefresh)
+	if err != nil {
+		runenv.Abort(err)
+		return
+	}
+
+	// Some time for the DHT to warm up
+	time.Sleep(time.Duration(20) * time.Second)
 
 	/// --- Act I
 
+	// TODO, only `p-providing` of the nodes provide a record and store its key on redis
+
+	var (
+		cidsToPublish     []cid.Cid
+		nCidsToPublish    = 5
+		cidsToNotPublish  []cid.Cid
+		nCidsToNotPublish = 5
+	)
+
+	for i := 0; i < nCidsToPublish; i++ {
+		v := fmt.Sprintf("%d -- gonna publish", i)
+		mhv := ipfsUtil.Hash([]byte(v))
+		cidsToPublish = append(cidsToPublish, cid.NewCidV0(mhv))
+	}
+
+	for i := 0; i < nCidsToNotPublish; i++ {
+		v := fmt.Sprintf("%d -- not gonna publish", i)
+		mhv := ipfsUtil.Hash([]byte(v))
+		cidsToNotPublish = append(cidsToNotPublish, cid.NewCidV0(mhv))
+	}
+
+	for i := 0; i < nCidsToPublish; i++ {
+		err := dht.Provide(ctx, cidsToPublish[i], true)
+		if err != nil {
+			runenv.Abort(fmt.Errorf("Failed on .Provide - %w", err))
```

Comment:
I'm always getting this error with:

```
<< 83046fff3822e0002d612cae80e53b0222c2ffb3127b193ff49e610503db556f >> {"runenv":{"test_plan":"dht","test_case":"find-providers","test_run":"aea7286e-07f5-4255-a19a-40b9319fd6f8","test_seq":1,"test_instance_count":5},"timestamp":1573212021273521900,"result":{"outcome":"aborted","reason":"Failed on .Provide - no known addresses for self. cannot put provider."}}
```

Which comes from https://github.com/libp2p/go-libp2p-kad-dht/blob/master/routing.go#L466-L471

Does https://github.com/libp2p/go-libp2p-kad-dht/blob/master/routing.go#L466-L471 that now by default, the DHT doesn't work on LANs?

Reply:
nvm. Was calling context cancel in the wrong place